src/libexpr/flake/flake.cc

```diff
@@ -570,19 +570,20 @@ LockedFlake lockFlake(
                             parents.push_back(*input.ref);
                             Finally cleanup([&]() { parents.pop_back(); });
 
-                            /* Recursively process the inputs of this
-                               flake. Also, unless we already have this flake
-                               in the top-level lock file, use this flake's
-                               own lock file. */
+                            /* Recursively process the inputs of this flake.
+                               Use this flake's own lock file; we don't want
+                               to leave the transitive inputs at the version
+                               associated with what this dependency used to be
+                               used to be */
+                            auto upstreamLock =
+                                LockFile::read(
+                                    inputFlake.sourceInfo->actualPath
+                                     + "/"
+                                     + inputFlake.lockedRef.subdir
+                                     + "/flake.lock").root.get_ptr(),
```

Comment:
And forgot to make this a semicolon:
```suggestion
                                     + "/flake.lock").root.get_ptr();
```